Enterprise coupons overview

mushtaqak opened the pull request:
The purpose of this PR is to add backend support for listing coupons for an enterprise needed for edx-portal.

From jira story:
> Returns a list of json objects with: Coupon ID (coupon product id), Coupon Name, Valid From and To, # of Codes, Usage Limitation (Maximum # of Usages per code), % of available Code redemptions available that have been redeemed.

Sample endpoint URL: https://ecommerce-entcoupon.sandbox.edx.org/api/v2/coupons/enterprise/cf246b88d5f64908a522fc307e0b0c60/overview/

Sample response:
```python
{
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
        {
            "end_date": "2019-03-12T00:00:00Z",
            "has_error": false,
            "id": 6,
            "max_uses": 2,
            "num_codes": 100,
            "num_unassigned": 0,
            "num_uses": 0,
            "start_date": "2018-11-19T00:00:00Z",
            "title": "Test edX coupon2",
            "usage_limitation": "Once per customer"
        },
        {
            "end_date": "2019-02-21T00:00:00Z",
            "has_error": false,
            "id": 5,
            "max_uses": null,
            "num_codes": 10,
            "num_unassigned": 0,
            "num_uses": 0,
            "start_date": "2018-11-19T00:00:00Z",
            "title": "Test coupon1",
            "usage_limitation": "Single use"
        }
    ]
}
```

[Backend API for Coupon Overview - ENT-1260](https://openedx.atlassian.net/browse/ENT-1260)

[Sandbox](https://ecommerce-entcoupon.sandbox.edx.org/coupons/)

**Test Instructions**
- [Create some coupons](https://ecommerce-entcoupon.sandbox.edx.org/coupons/new/) and associate it with some client (eg: entX).
> After you have created a couple of coupons with `entX` client, You might want to update `client`'s `enterprise_customer_uuid` (eg: cf246b88d5f64908a522fc307e0b0c65) from [django admin](https://ecommerce-entcoupon.sandbox.edx.org/admin/core/businessclient/), since sandbox is not linked with LMS.
- [Observe all coupons](https://ecommerce-entcoupon.sandbox.edx.org/coupons/) list in ecommerce.
- Observe the list of coupons associated with <enterprise_customer_id> at 
https://ecommerce-business.sandbox.edx.org/api/v2/coupons/enterprise/cf246b88d5f64908a522fc307e0b0c65/overview/

mushtaqak commented:
@brittneyexline Can you have a very quick look at the following sample API response and see if are going in correct direction.
```python
{
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
        {
            "end_date": "2019-03-08T00:00:00Z",
            "id": 2,
            "max_uses": null,
            "num_uses": 0,
            "quantity": 1,
            "start_date": "2018-11-12T00:00:00Z",
            "title": "TESt2",
            "voucher_type": "Single use"
        },
        {
            "end_date": "2019-02-01T00:00:00Z",
            "id": 1,
            "max_uses": null,
            "num_uses": 0,
            "quantity": 100,
            "start_date": "2018-11-12T00:00:00Z",
            "title": "TEST_COUPON",
            "voucher_type": "Multi-use"
        }
    ]
}
```
Esp. about `% of available Code redemptions available that have been redeemed.` part mentioned in the PR description. I mean, we don't need to calculate percentage here, we just need to send the the data to front-end, right?

adamstankiewicz commented:
> Esp. about `% of available Code redemptions available that have been redeemed.` part mentioned in the PR description. I mean, we don't need to calculate percentage here, we just need to send the the data to front-end, right?

Yeah, the front-end will calculate the percentage of redemptions used/available.

adamstankiewicz commented:
@mushtaqak @brittneyexline Will we be able to know through this API whether any of the codes have an error (e.g., undeliverable email)? The front-end has an assumption that the API response for the coupons will contain a boolean field like `has_error` to determine whether we should show an error state for a coupon.

adamstankiewicz commented:
@mushtaqak In the above example API response, do any of those fields align with the number of unassigned codes remaining? Is it the `quantity` field?

See the below screenshot (i.e., "Unassigned Codes") where this data will be used:

![image](https://user-images.githubusercontent.com/2828721/48428394-ecb0e180-e738-11e8-8c0b-c5d126283cf9.png)

brittneyexline commented:
@adamstankiewicz i think including the error at this level is a question of what the ui needs. however, it probably won't be implemented with this change since we haven't implemented notifying or assigning users yet.

robrap commented:
@adamstankiewicz @mushtaqak: It would be great if names can be updated to be more readily understandable.  So if it is `quantity`, could we rename to `num_unassigned` (or something like that)?

adamstankiewicz commented:
> @adamstankiewicz i think including the error at this level is a question of what the ui needs. however, it probably won't be implemented with this change since we haven't implemented notifying or assigning users yet.

Yeah, that makes sense. Just making sure it can eventually be supported in this API 😄

adamstankiewicz commented:
> @adamstankiewicz @mushtaqak: It would be great if names can be updated to be more readily understandable. So if it is `quantity`, could we rename to `num_unassigned` (or something like that)?

Agreed 👍

mushtaqak commented:
@adamstankiewicz `quantity` is `# of codes` mentioned in the ticket. However, I think the mockups has changed, so we can either remove it or UI won't use it (atleast not in the overview screen) 

About, `Unassigned Codes`, it's functionality part is not implemented yet (scoped in some other JIRA story). So, I will add just the placeholder in this PR.

Regarding error status, can you explain the a bit more, what info front end would need? copoun status? Sorry, I don't have much of info on coupon/code terms.

adamstankiewicz commented:
@mushtaqak Sounds good about the placeholder for number of unassigned codes. If we do keep the `quantity` field, can we rename it to be more explicit about what it's referring to?

As for the error status, the UI needs to know whether a code(s) within a coupon contains an error to properly prompt the user that an error exists. The coupon doesn't need to know what the actual error is; just that one exists. It could just be a boolean field in the API response, like `has_error: true`. 

For reference, this is the coupon error state in the UI:

![image](https://user-images.githubusercontent.com/2828721/48433138-5da9c680-e744-11e8-87de-79505e8654e4.png)

mushtaqak commented:
@brittneyexline @adamstankiewicz Can you guys review this?

brittneyexline commented on ecommerce/extensions/api/v2/views/enterprise.py:
```diff
@@ -48,7 +53,10 @@ class EnterpriseCouponViewSet(CouponViewSet):
     def get_queryset(self):
+        enterprise_id = self.kwargs.get('enterprise_id')
         invoices = Invoice.objects.filter(business_client__enterprise_customer_uuid__isnull=False)
+        if enterprise_id:
```
nit: more readable if this is an if/else statement

brittneyexline commented on ecommerce/extensions/api/v2/views/enterprise.py:
```diff
@@ -60,8 +68,31 @@ def get_queryset(self):
+        """
+        Overview of Enterprise coupons.
+        Returns the following data:
+            - Copoun ID
```
nit: typo (Coupon)

brittneyexline commented on ecommerce/extensions/api/v2/tests/views/test_enterprise.py:
```diff
@@ -63,3 +81,83 @@ def test_get_customers(self, mock_client):
+    Tests EnterpriseCouponViewSet works as expected.
+    """
+
+    def setUp(self):
```
you'll have to rebase on my branch again - i have just added a bunch of tests so some of this is redundant

brittneyexline commented on ecommerce/extensions/api/v2/tests/views/test_enterprise.py:
```diff
@@ -114,8 +115,40 @@ def setUp(self):
+            'title': 'Test coupon 3'
+        })
+
+        self.coupon3 = self.create_get_coupon(data3)
```
i'd prefer if we move these coupon creations to the test method that needs them.

brittneyexline commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -618,6 +618,48 @@ class Meta(object):
+
+    # Redemption count.
+    def get_num_uses(self, obj):
+        offer = retrieve_offer(obj)
```
we will need to revisit this because many vouchers will have two offers, and we'll probably want to add together the num_applications on each offer to get the actual total number of applications.

mushtaqak replied:
@brittneyexline Do you think, this will be covered in this PR? What do we have to do? I looked at a couple of places in codebase, do you think something like this would be needed?

```python
def redemptions(self, obj):
        voucher = getattr(obj, 'voucher', obj)
        offer = voucher.best_offer
        redemption_count = offer.num_applications

        if voucher.usage == Voucher.SINGLE_USE:
            max_coupon_usage = 1
            redemption_count = voucher.num_orders
        elif voucher.usage != Voucher.SINGLE_USE and offer.max_global_applications is None:
            max_coupon_usage = 10000
        else:
            max_coupon_usage = offer.max_global_applications

        return {
            'used': redemption_count,
            'available': max_coupon_usage,
        }
```

brittneyexline replied:
for now let's leave it as is... i haven't quite figured out how we're going to manage these things when we have to switch which offer we use

brittneyexline replied:
after pulling some data for an audit of the coupons yesterday, i remembered that there is a `num_orders` field on the voucher - i think using this value will give us the best value for num_uses without having to account for the multiple conditional offers.

mushtaqak commented:
@brittneyexline I have addressed all of your feedback except the `num_uses` where I have asked the question. Rest this is ready for your review again.

mushtaqak commented:
@georgebabey Can you review this as well ?

adamstankiewicz commented:
@mushtaqak @brittneyexline Just taking a look at the API response, and all the data the frontend needs seems to be there! Nice 👍

adamstankiewicz commented:
It looks like the API response is paginated (e.g., has `prev` and `next` fields), but it doesn't have a `num_pages` field like API responses in our other services. Is this expected?

mushtaqak commented:
@adamstankiewicz I guess REST APIs have a `count` field (not the`num_pages`). Below is the response containing fields returned ...
```python
{
    "count": ...,
    "next": ...,
    "previous": ...,
    "results": [...]
}

brittneyexline commented on ecommerce/extensions/api/v2/urls.py:
```diff
@@ -105,6 +105,8 @@
     .register(r'products', product_views.ProductViewSet, base_name='catalog-product',
               parents_query_lookups=['stockrecords__catalogs'])
 router.register(r'coupons', coupon_views.CouponViewSet, base_name='coupons')
+router.register(r'coupons/enterprise/(?P<enterprise_id>.+)', enterprise_views.EnterpriseCouponViewSet,
```
should we add /overview to the pattern so that we match exactly?

brittneyexline replied:
i noticed that the paths are slightly different between yours and Ammar's PR - here it is `coupons/enterprise` and on his and for the other endpoints it is `enterprise/coupons`. i think you should make sure the path is consistent with the other endpoints.

georgebabey commented:
@mushtaqak Looks like this is ready to go but theres some conflicts to resolve. Can you update this and I'll get it merged in.

georgebabey commented on ecommerce/extensions/api/v2/tests/views/test_enterprise.py:
```diff
@@ -63,3 +88,309 @@ def test_get_customers(self, mock_client):
+
+        overview_response = self.get_response_json(
+            'GET',
+            reverse('api:v2:enterprise-coupons-(?P<enterprise-id>.+)/overview-list',
```
Are you sure this is correct? It looks like it should be `/overview` instead of `/overview-list` https://github.com/edx/ecommerce/pull/1974/commits/91fcfa83978643ebb6d3dd20bf87016f2746750b#diff-141fa9c110c00aba0d7f70839c4a77bbR179

mushtaqak commented:
@georgebabey This is ready to merge!
